susemanager-utils/susemanager-sls/src/modules/mgractionchains.py

```diff
@@ -178,27 +209,47 @@ def resume():
 
     This method is called by the Salt Reactor as a response to the 'minion/start/event'.
     '''
-    next_chunk = _read_next_ac_chunk()
-    if not next_chunk:
+    yaml_file = _read_next_ac_chunk()
+    if not yaml_file:
         return {}
-    if type(next_chunk) != dict:
+    if type(yaml_file) != dict:
         err_str = "Not able to resume Action Chain execution! Malformed " \
-                  "'_mgractionchains.conf' found: {0}".format(next_chunk)
+                  "'_mgractionchains.conf' found: {0}".format(yaml_file)
         log.error(err_str)
         raise CommandExecutionError(err_str)
-    next_chunk = next_chunk.get('next_chunk')
+    next_chunk = yaml_file.get('next_chunk')
     log.debug("Resuming execution of SUSE Manager Action Chain -> Target SLS: "
               "{0}".format(next_chunk))
-
-    inside_transaction = os.environ.get("TRANSACTIONAL_UPDATE")
-    if __grains__.get("transactional") and not inside_transaction:
-        return __salt__['transactional_update.sls'](next_chunk, queue=True)
+    
+    transactional_update = __grains__.get("transactional")
+    reboot_required = False
+    inside_transaction = False
+    if transactional_update:
+        reboot_required = yaml_file.get('reboot_required')
+        inside_transaction = os.environ.get("TRANSACTIONAL_UPDATE")
+
+    if transactional_update and reboot_required:
+        ret =  __salt__['transactional_update.sls'](next_chunk, queue=True, activate_transaction=True)
+    elif transactional_update and not inside_transaction:
+        ret =  __salt__['transactional_update.sls'](next_chunk, queue=True)
     else:
-        return __salt__['state.sls'](next_chunk, queue=True)
+        ret = __salt__['state.sls'](next_chunk, queue=True)
+
+    if isinstance(ret, list):
+        raise CommandExecutionError(ret)
+    return ret
 
-def clean():
+def clean(actionchain_id=None, current_action_id=None, reboot_required=None):
     '''
     Clean execution of an Action Chain by removing '_mgractionchains.conf'.
     '''
     _read_next_ac_chunk()
-    return {"success": True}
+    yaml_dict = {}
+    yaml_dict['success'] = True
+    if actionchain_id:
+        yaml_dict['actionchain_id'] = actionchain_id
+    if current_action_id:
+        yaml_dict['current_action_id'] = current_action_id
+    if reboot_required:
+        yaml_dict['reboot_required'] = reboot_required
+    return yaml_dict
```

Comment:
It surprised me that a function called `clean` returns so much data, how is it expected to be used? It looks like it's only returning its input arguments.

Reply:
actually there's a reason for it. These parameters were introduced to fix
> although reboot is performed, WebUI events -> pending still show that the reboot action was not executed.

In minion without transactional update, the job result for the reboot action was collected by `SaltServerActionService::handleActionChainResult` : if reboot was scheduled successfully, `handleActionChainResult` put the action in `picked up` status. When the reboot completed, the `picked up` status become `successful`.

In minion with transactional update, there's no job result for the reboot action (since there's no action, the reboot is applied by `activate_transaction=true`), so I changed `mgractionchains.next` and  `mgractionchains.clean`  in order to provide as job result all the information required by `SaltServerActionService::handleActionChainResult` to set a "fake" reboot action as `picked up` .  You can find the code here: https://github.com/uyuni-project/uyuni/blob/c0588c3fe46f62b0a6af4bd7aad6538694119fb8/java/code/src/com/suse/manager/webui/services/SaltServerActionService.java#L2805

(actually `actionchain_id` is not used, just `current_action_id` and `reboot_required`, but I prefer to add this one because I think it could help if error happens)